Islandora-CLAW/Crayfish
# Issue 612

**GitHub Issue**: Resolves Islandora-CLAW/CLAW#612

* Blocked by ~~https://github.com/fcrepo4-labs/fcrepo-api-x/pull/122~~ ~~https://github.com/Islandora-CLAW/CLAW/issues/666~~

# What does this Pull Request do?

Adds OPTIONS requests to Houdini and Hypercube so they can be added to Api-X

# What's new?
OPTIONS requests for each operation, plus tests.  The requests just serve static ttl files describing the services.

# How should this be tested?

Pull in the code for this PR and run `composer update` on both Houdini and Hypercube.

After https://github.com/fcrepo4-labs/fcrepo-api-x/pull/122 lands, you'll need to load the services into Api-X using its loader service.

- `curl -X POST  -H "Authorization: Bearer islandora" -H "Content-Type: text/plain" -d "http://localhost:8000/houdini/convert" "http://localhost:8081/services//apix:load"`
- `curl -X POST  -H "Authorization: Bearer islandora" -H "Content-Type: text/plain" -d "http://localhost:8000/houdini/identify" "http://localhost:8081/services//apix:load"`
- `curl -X POST  -H "Authorization: Bearer islandora" -H "Content-Type: text/plain" -d "http://localhost:8000/hypercube" "http://localhost:8081/services//apix:load"`
- Then, ingest a tiff into Fedora.  Mine was located at `http://localhost:8080/fcrepo/rest/test.tiff`
- Perform a `HEAD` request against the tiff through the Api-X proxy: `curl -I -H "Authorization: Bearer islandora" http://localhost:8081/fcrepo/rest/test.tiff`
- In the returned headers, you should see a link to discover the services for the tiff.  Mine looked like `Link: <http://localhost:8081/discovery/test.tiff>; rel="service"`
- `GET` the discovery document: ` curl -H "Authorization: Bearer islandora" http://localhost:8081/discovery/test.tiff`
- It should return rdf describing the available services.  Mine looked like this:
```
<>      a       <http://fedora.info/definitions/v4/service#ServiceDocument> ;
        <http://fedora.info/definitions/v4/service#isServiceDocumentFor>
                <http://localhost:8081/fcrepo/rest/test.tiff> ;
        <http://www.openarchives.org/ore/terms/describes>
                <#services> .

<#services>  <http://www.openarchives.org/ore/terms/aggregates>
                <#2c1aa6b6-4bc6-451a-b3f0-80332e1613e8> , <#5335cf4f-c041-45a0-98e9-5a48eaa56d69> , <#229cc418-620b-4a9e-8bc8-60fee5b648ae> .

<#229cc418-620b-4a9e-8bc8-60fee5b648ae>
        a       <http://fedora.info/definitions/v4/service#ServiceInstance> ;
        <http://fedora.info/definitions/v4/service#hasEndpoint>
                <http://localhost:8081/services/test.tiff/svc:convert> ;
        <http://fedora.info/definitions/v4/service#isFunctionOf>
                <http://localhost:8081/fcrepo/rest/test.tiff> ;
        <http://fedora.info/definitions/v4/service#isServiceInstanceOf>
                <http://islandora.ca/CLAW#ConvertService> ;
        <http://fedora.info/definitions/v4/service#serviceInstanceExposedBy>
                <http://localhost:8081/fcrepo/rest/test.tiff> .

<#5335cf4f-c041-45a0-98e9-5a48eaa56d69>
        a       <http://fedora.info/definitions/v4/service#ServiceInstance> ;
        <http://fedora.info/definitions/v4/service#hasEndpoint>
                <http://localhost:8081/services/test.tiff/svc:identify> ;
        <http://fedora.info/definitions/v4/service#isFunctionOf>
                <http://localhost:8081/fcrepo/rest/test.tiff> ;
        <http://fedora.info/definitions/v4/service#isServiceInstanceOf>
                <http://islandora.ca/CLAW#IdentifyService> ;
        <http://fedora.info/definitions/v4/service#serviceInstanceExposedBy>
                <http://localhost:8081/fcrepo/rest/test.tiff> .

<#2c1aa6b6-4bc6-451a-b3f0-80332e1613e8>
        a       <http://fedora.info/definitions/v4/service#ServiceInstance> ;
        <http://fedora.info/definitions/v4/service#hasEndpoint>
                <http://localhost:8081/services/test.tiff/svc:ocr> ;
        <http://fedora.info/definitions/v4/service#isFunctionOf>
                <http://localhost:8081/fcrepo/rest/test.tiff> ;
        <http://fedora.info/definitions/v4/service#isServiceInstanceOf>
                <http://islandora.ca/CLAW#OcrService> ;
        <http://fedora.info/definitions/v4/service#serviceInstanceExposedBy>
                <http://localhost:8081/fcrepo/rest/test.tiff> .
```
- As you can see, all three services are available should be available through Api-X.  Check out the objects of the `http://fedora.info/definitions/v4/service#hasEndpoint` triples to see where you can issue an HTTP request.
- `curl -H "Authorization: Bearer islandora" http://localhost:8081/services/test.tiff/svc:identify` should return the json from the identify command.
- `curl -H "Authorization: Bearer islandora" http://localhost:8081/services/test.tiff/svc:ocr` should return the OCR output
- `curl -H "Authorization: Bearer islandora" -H "Accept: image/png" http://localhost:8081/services/test.tiff/svc:convert` should return a png version of the image.

# Interested parties
@Islandora-CLAW/committers @ajs6f @birkland
